RFC: Split TypeMaps with Variance

diesalbla opened the pull request:
This PR introduces some "refactors" in the `TypeMap` hierarchy. We noted that the "trackVariance" attribute of a type map was never changed, which suggests it is a static element, more represented through types, than through a field. In particular, the changes are: 

- We remove from the `TypeMap` abstract class the `trackVariance` field, and all the methods that were specifically tied to that. 
- We introduce an abstract `VariancedTypeMap` class, for those `TypeMap`s that need to track variance. We add here the fields and methods we cut from `TypeMap`. 
- We add a companion object with an `unapply` method. We use it at several places in which, before, there was by default a called to the `flipped`. 

I have tested this changes with `partest`. Almost all of the tests pass ok, except for two. 
- One is related to the Java Universe of  `scala.reflect`, which now includes `VariancedTypeMap`. 
- The other one is [`test/files/neg/t0764.scala`](https://github.com/scala/scala/blob/2.13.x/test/files/neg/t0764.scala), for which the expectation is that it would not compile, but which it does in this branch. Interestingly, the comments by @adriaanm suggests that it should actually be a positive example, and that this branch may actually be fixing an old bug. The compilation for the larger file, `t0764b.scala`  does not change.

@adriaanm Could you please weigh in, as to whether being able to compile that file is either break or a bug-fix?

diesalbla committed:
Split TypeMaps with Variance

We noted that the "trackVariance" attribute of a type map was never
changed, which suggests that it was a static element better expressed
through a type than through a boolean field.

We modify the TypeMap class Hierarchy, so that type maps that do need
to track variance have their own specific  class "VariancedTypeMap".
We move the variance-handling methods, and branches of methods, there.

We add a companion object to VariancedTypeMap, which takes care of
checking if a TypeMap is a VariancedTypeMap or not.

This code seems to break the execution in partest of one test,
for which we are not sure if it is actually fixing a bug.

adriaanm commented on src/reflect/scala/reflect/internal/tpe/TypeMaps.scala:
```diff
@@ -209,6 +183,45 @@ private[internal] trait TypeMaps {
+
+  object VariancedTypeMap {
+    def unapply(tm: TypeMap): Option[VariancedTypeMap] =
+      if (tm.isInstanceOf[VariancedTypeMap]) Some(tm.asInstanceOf[VariancedTypeMap]) else None
```
Why prefer `case VariancedTypeMap(vtm) =>` over `case vtm: VariancedTypeMap =>`? The latter seems fine & simpler to me?

(Side note, you could define this more succinctly as `def unapply(vtm: VariancedTypeMap): Some[VariancedTypeMap] = Some(vtm)`. The pattern matcher will insert the instance-of checks for you before calling the extractor.)

diesalbla replied:
Using the `case vtm: VariancedTypeMap` is cleaner, so I have removed this object. This also allows me to revert the change done for the `JavaUniverseForce`.

diesalbla committed:
Attend code review comments: remove VariancedTypeMap object.

We can match the subtype by means of the `:` check.

diesalbla committed:
Further changes

We repair a bug in the previous commits: the `applyToSymbolInfo`
method had to be redefined to account for variance.
We remove the `mapOverArgs` method from the TypeMap trait, since
for non-varianced maps it is equivalent to `mapConserve`.

diesalbla commented:
The failing test was actually due to an error: when splitting the `VariancedTypeMap`, one private method from `TypeMap` that was adapted to variance was not overridden in `VariancedTypeMap`. 

I am closing this PR to re-submit the changes elsewhere.
